Parametric/Denotation/CachingEvaluation.agda

```diff
@@ -0,0 +1,198 @@
+------------------------------------------------------------------------
+-- INCREMENTAL λ-CALCULUS
+--
+-- Caching evaluation
+------------------------------------------------------------------------
+
+import Parametric.Syntax.Type as Type
+import Parametric.Syntax.Term as Term
+
+import Parametric.Syntax.MType as MType
+import Parametric.Syntax.MTerm as MTerm
+
+import Parametric.Denotation.Value as Value
+import Parametric.Denotation.Evaluation as Evaluation
+import Parametric.Denotation.MValue as MValue
+import Parametric.Denotation.CachingMValue as CachingMValue
+import Parametric.Denotation.MEvaluation as MEvaluation
+
+module Parametric.Denotation.CachingEvaluation
+    {Base : Type.Structure}
+    (Const : Term.Structure Base)
+    (⟦_⟧Base : Value.Structure Base)
+    (⟦_⟧Const : Evaluation.Structure Const ⟦_⟧Base)
+    (ValConst : MTerm.ValConstStructure Const)
+    (CompConst : MTerm.CompConstStructure Const)
+    -- I should really switch to records - can it get sillier than this?
```

Comment:
More precisely, this is the kind of thing ML functors are designed to replace. They have also subtyping though.